### PR TITLE
Adds the charm-pull-source command

### PR DESCRIPTION
**Pull request by tvansteenburgh**
TLDR, run `charm-pull-source -h`.

Downloads the source code for a charm, layer, or interface.

The "thing-to-download" can be specified using any of the following forms:

 - [cs:]charm
 - [cs:]series/charm
 - [cs:]~user/charm
 - [cs:]~user/series/charm
 - layer:layer-name
 - interface:layer-name

If a download directory is not specified, the following environment vars
will be used to determine the download location:

 - For charms, $JUJU_REPOSITORY
 - For layers, $LAYER_PATH
 - For interfaces, $INTERFACE_PATH

If a download location can not be determined from environment variables,
the current working directory will be used.

The download is aborted if the destination directory already exists.

### CODE REVIEW

**Comment by tvansteenburgh** [comment]
Fixes #121

**Comment by marcoceppi** [comment]
User experience is pretty prickly.

StackTrace is path exists
```
(py27) marco@T430:~/Projects/charm-tools$ charm pull-source layer:django
Traceback (most recent call last):
  File "/home/marco/Projects/charm-tools/.tox/py27/bin/charm-pull-source", line 9, in <module>
    load_entry_point('charm-tools', 'console_scripts', 'charm-pull-source')()
  File "/home/marco/Projects/charm-tools/charmtools/pullsource.py", line 141, in main
    return download_item(args.item, args.dir)
  File "/home/marco/Projects/charm-tools/charmtools/pullsource.py", line 97, in download_item
    dest = fetcher.fetch(dir_)
  File "/home/marco/Projects/charm-tools/charmtools/build/fetchers.py", line 70, in fetch
    return super(InterfaceFetcher, self).fetch(dir_)
  File "/home/marco/Projects/charm-tools/charmtools/fetchers.py", line 231, in fetch
    shutil.copytree(self.path, dst, symlinks=True)
  File "/usr/lib/python2.7/shutil.py", line 177, in copytree
    os.makedirs(dst)
  File "/home/marco/Projects/charm-tools/.tox/py27/lib/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 17] File exists: '/home/marco/Projects/charms/layers/django'
```

Stack trace is layer doesn't exist
```
(py27) marco@T430:~/Projects/charm-tools$ charm pull-source layer:lts
INFO: Starting new HTTP connection (1): interfaces.juju.solutions
Traceback (most recent call last):
  File "/home/marco/Projects/charm-tools/.tox/py27/bin/charm-pull-source", line 9, in <module>
    load_entry_point('charm-tools', 'console_scripts', 'charm-pull-source')()
  File "/home/marco/Projects/charm-tools/charmtools/pullsource.py", line 141, in main
    return download_item(args.item, args.dir)
  File "/home/marco/Projects/charm-tools/charmtools/pullsource.py", line 96, in download_item
    fetcher = fetchers.get_fetcher(item)
  File "/home/marco/Projects/charm-tools/charmtools/fetchers.py", line 375, in get_fetcher
    raise FetchError('No fetcher for url: %s' % url)
charmtools.fetchers.FetchError: No fetcher for url: layer:lts
```

And I can't pull a valid layer
```
(py27) marco@T430:~/Projects/charm-tools$ charm pull-source layer:ansible-base
INFO: Starting new HTTP connection (1): interfaces.juju.solutions
Traceback (most recent call last):
  File "/home/marco/Projects/charm-tools/.tox/py27/bin/charm-pull-source", line 9, in <module>
    load_entry_point('charm-tools', 'console_scripts', 'charm-pull-source')()
  File "/home/marco/Projects/charm-tools/charmtools/pullsource.py", line 141, in main
    return download_item(args.item, args.dir)
  File "/home/marco/Projects/charm-tools/charmtools/pullsource.py", line 97, in download_item
    dest = fetcher.fetch(dir_)
  File "/home/marco/Projects/charm-tools/charmtools/build/fetchers.py", line 80, in fetch
    target = dir_ / basename
TypeError: unsupported operand type(s) for /: 'str' and 'Path'
```

default logging is too VERBOSE
```
(py27) marco@T430:~/Projects/charm-tools$ charm pull-source mysql
INFO: Starting new HTTPS connection (1): api.jujucharms.com
Downloaded cs:mysql to /home/marco/Projects/charms/mysql
```

Otherwise, :+1:

**Comment by marcoceppi** [comment]
Charm fetcher should attempt to read vcs information first from extra-info in v4 api:

https://api.jujucharms.com/charmstore/v4/trusty/mysql/meta/any?include=extra-info

For charms uploaded directly to the store this section will be present but empty

https://api.jujucharms.com/charmstore/v4/~marcoceppi/trusty/ubucon/meta/any?include=extra-info

**Comment by tvansteenburgh** [comment]
@marcoceppi I believe this addresses all the issues you raised. Thanks for the review!

**Comment by marcoceppi** [comment]
LGTM +1 :boat:
